beacon/light/committee_chain.go

```diff
@@ -0,0 +1,555 @@
+	var (
+		slotTime = int64(time.Second) * int64(s.config.GenesisTime+head.Header.Slot*12)
+		age      = time.Duration(s.unixNano() - slotTime)
+	)
+	if s.enforceTime && age < 0 {
+		return false, age
+	}
+	committee := s.getSyncCommittee(types.SyncPeriod(head.SignatureSlot))
+	if committee == nil {
+		return false, age
+	}
+	if signingRoot, err := s.config.Forks.SigningRoot(head.Header); err == nil {
+		return s.sigVerifier.verifySignature(committee, signingRoot, &head.Signature), age
+	}
+	return false, age
+}
+
+// verifyUpdate checks whether the header signature is correct and the update
+// fits into the specified constraints (assumes that the update has been
+// successfully validated previously)
+func (s *CommitteeChain) verifyUpdate(update *types.LightClientUpdate) bool {
+	// Note: SignatureSlot determines the sync period of the committee used for signature
+	// verification. Though in reality SignatureSlot is always bigger than update.Header.Slot,
+	// setting them as equal here enforces the rule that they have to be in the same sync
+	// period in order for the light client update proof to be meaningful.
+	ok, age := s.verifySignedHeader(update.AttestedHeader)
+	if age < 0 {
+		log.Warn("Future committee update received", "age", age)
+	}
+	return ok
+}
+
+// canonicalStore stores instances of the given type in a database and caches
+// them in memory, associated with a continuous range of period numbers.
+type canonicalStore[T any] struct {
+	Range
+	db        ethdb.KeyValueStore
+	keyPrefix []byte
+	cache     *lru.Cache[uint64, T]
+	encode    func(T) ([]byte, error)
+	decode    func([]byte) (T, error)
+}
+
+// newCanonicalStore creates a new canonicalStore.
+func newCanonicalStore[T any](db ethdb.KeyValueStore, keyPrefix []byte,
+	encode func(T) ([]byte, error), decode func([]byte) (T, error)) *canonicalStore[T] {
+	cs := &canonicalStore[T]{
+		db:        db,
+		keyPrefix: keyPrefix,
+		encode:    encode,
+		decode:    decode,
+		cache:     lru.NewCache[uint64, T](100),
+	}
+	var (
+		iter = db.NewIterator(keyPrefix, nil)
+		kl   = len(keyPrefix)
+	)
+	for iter.Next() {
+		period := binary.BigEndian.Uint64(iter.Key()[kl : kl+8])
+		if cs.First == 0 {
+			cs.First = period
+		} else if cs.AfterLast != period {
+			if iter.Next() {
+				log.Error("Gap in the canonical chain database")
+			}
+			break // continuity guaranteed
+		}
+		cs.AfterLast = period + 1
+	}
+	iter.Release()
+	return cs
+}
+
+// getDbKey returns the database key belonging to the given period.
+func (cs *canonicalStore[T]) getDbKey(period uint64) []byte {
+	var (
+		kl  = len(cs.keyPrefix)
+		key = make([]byte, kl+8)
+	)
+	copy(key[:kl], cs.keyPrefix)
+	binary.BigEndian.PutUint64(key[kl:], period)
+	return key
+}
+
+// add adds the given item to the database. It also ensures that the range remains
+// continuous. Can be used both in batch mode and as a standalone operation.
+func (cs *canonicalStore[T]) add(batch ethdb.Batch, period uint64, value T) {
```

Comment:
Done, now the error logs are still printed in `add` (because these scenarios should never happen) but an error is also returned and bubbled up by the callers.